File: C++/Knapsack.cpp
#include <iostream>
#include <vector>
using namespace std;

#define ll long long 

ll N, W;
vector<ll> Ws;
vector<ll> Vs;
vector<vector<ll>> F;

ll MFKsnapsack(int i, int j){
    if(F[i][j] < 0){
        ll value;
        if(j < Ws[i]){
            value = MFKsnapsack(i-1, j);
        } else {
            value = max(MFKsnapsack(i-1, j), Vs[i]+MFKsnapsack(i-1, j-Ws[i]));
        }
        F[i][j] = value;
    }
    return F[i][j];
}

int main(){
    cin >> N >> W;

    Ws.push_back(0);
    Vs.push_back(0);

    for(int i=1; i<N+1; i++){
        ll x, y;
        cin >> x >> y;
        Ws.push_back(x), Vs.push_back(y);
    }

    // Constroi tabela
    for(int i=0; i<Vs.size(); i++){
        vector<ll> line;

        for(int j=0; j<W+1; j++){
            if(i == 0 || j == 0){
                line.push_back(0);
            }
            else {
                line.push_back(-1);
            }
        }
        F.push_back(line);
    }
    
    cout << MFKsnapsack(N, W) << endl;
}

File: Python/Outros/Knapsack.py
def MFKsnapsack(i, j):
    if F[i][j] < 0:
        if j < Ws[i]:
            value = MFKsnapsack(i-1, j)
        else:
            value = max(MFKsnapsack(i-1, j), Vs[i]+MFKsnapsack(i-1, j-Ws[i]))

        F[i][j] = value

    return F[i][j]

N, W = list(map(int, input().split()))
Ws = [0] * (N+1)
Vs = [0] * (N+1)

for i in range(1, N+1):
    Ws[i], Vs[i] = list(map(int, input().split()))

# Constroi tabela
F = []
for i in range(len(Vs)):
    line = []
    for j in range(W+1):
        if i == 0 or j == 0:
            line.append(0)
        else:
            line.append(-1)
    F.append(line)

print(MFKsnapsack(N, W))